crates/subspace-transaction-pool/src/bundle_validator.rs

```diff
@@ -0,0 +1,256 @@
+            self.bundle_stored_in_last_k.push_front(block_bundles);
+        }
+        Ok(())
+    }
+
+    /// Collect bundles within the new blocks of the best chain, blocks are handled from
+    /// older blcok to newer blcok, an `Err` may return in the middle and left some blocks
+    /// unhandled, these blocks will be handled when processing the next new best block.
+    fn on_import_new_best_block(
+        &mut self,
+        new_best_hash: Block::Hash,
+    ) -> Result<(), sp_blockchain::Error> {
+        let current_best_hash = match self.best_hash() {
+            Some(h) => h,
+            // Collect bundle from the last K blocks if it have not process any block
+            None => {
+                self.collect_bundle_from_last_k(new_best_hash)?;
+                return Ok(());
+            }
+        };
+
+        let route = sp_blockchain::tree_route(&*self.client, current_best_hash, new_best_hash)?;
+        let (retracted, enacted) = (route.retracted(), route.enacted());
+
+        // Remove bundles from the stale fork
+        for retracted_block in retracted {
+            match self.bundle_stored_in_last_k.front() {
+                Some((block_hash, _)) if *block_hash == retracted_block.hash => {
+                    self.bundle_stored_in_last_k.pop_front();
+                }
+                bb => {
+                    return Err(sp_blockchain::Error::Application(Box::from(
+                        format!(
+                            "Got wrong block from the bundle-collector, expect {:?}, got {:?}, this should not happen",
+                            retracted_block,
+                            bb.map(|(block_hash, _)| block_hash),
+                        ),
+                    )));
+                }
+            }
+        }
+
+        // Add bundles from the new block of the best fork
+        for enacted_block in enacted {
+            let block_bundles = self.collect_bundle_from_block(enacted_block.hash)?;
+            self.bundle_stored_in_last_k.push_front(block_bundles);
+        }
+
+        // Remove blocks from the back end to keep at most the bundle of the last K blocks
+        self.bundle_stored_in_last_k.truncate(self.confirm_depth_k);
+
+        Ok(())
+    }
+
+    pub async fn update_recent_stored_bundles(&mut self, new_best_hash: Block::Hash) {
+        let pre_best_hash = self.best_hash();
+        if let Err(err) = self.on_import_new_best_block(new_best_hash) {
+            tracing::error!(
+                %err,
+                "Failed to handle new best block for bundle-collector"
+            );
+        }
+        if pre_best_hash != self.best_hash() {
+            self.bundle_syncer
+                .update(self.bundle_stored_in_last_k.clone());
+        }
+    }
+}
+
+/// `BundleStoredInLastK` used to share the bundles of the last K blocks with different thread
+#[derive(Clone)]
+struct BundleStoredInLastK<Block: BlockT> {
```

Comment:
I don't understand why `BundleCollector` (and this is not a good name) is even needed. Wouldn't just a sharable `BundleStoredInLastK` be simpler?

```rust
struct BundleStoredInLastK {
    client: Arc<Client>,
    bundle_stored_in_last_k: Arc<Mutex<VecDeque<Block::Hash, HashSet<Hash>>>>,
    confirmation_depth_k: usize,
}

impl BundleStoredInLastK {
    pub fn new(client: Arc<Client>) -> Self {
    }

    pub async fn on_new_best_block(&self) {
        // Update `bundle_stored_in_last_k`
    }

    pub fn is_duplicated_bundle(&self, bundle_hash: Hash) -> bool {
        // Return whether `bundle_hash` exists in `bundle_stored_in_last_k`
    }
}

struct BundleValidator {
    client,
    bundle_stored_in_last_k: BundleStoredInLastK
}

impl BundleValidator {
    pub fn new(client: Client, bundle_stored_in_last_k: BundleStoredInLastK) -> Self {
    }
}

impl BundleValidator {
    fn validate_bundle(&self, bundle) {
        // Check the duplicated bundle.
        let bundle_hash = bundle.hash();
        if self.bundle_stored_in_last_k.is_duplicated_bundle(bundle_hash) {
            return Err(BundleError::Duplicated)
        }

        // Check the receipts
    }
}
```

Reply:
The `BundleStoredInLastK` you proposed here almost like identical to `BundleCollector`, so except the naming I guess your question here would be: why in current implement there are both `bundle_syncer` and `bundle_stored_in_last_k`
```rust
bundle_syncer: BundleStoredInLastK<Block>, // namely `Arc<Mutex<VecDeque<Block::Hash, HashSet<Hash>>>>`
bundle_stored_in_last_k: VecDeque<(Block::Hash, HashSet<Hash>)>,
```
instead of just `bundle_syncer` and make `BundleCollector` shared between thread?

The reason is while maintaining `bundle_stored_in_last_k` in `on_import_new_best_block` we need to access this filed quite many time and in between there may include some database reading, we do not want to doing that while holding a `Mutex` as the worker is running inside `tokio` and it will also blocking txn validation.

We can also twist the logic in `on_import_new_best_block` to make it avoid database reading while holding the `Mutex`, but that will hurt its readability and maintainability which is more important I think.

Reply:
I'm unsure we should take into account the database operations here because a) the transactions are validated concurrently, it has no influence on the other transactions except `submit_bundle`, and I think we probably ought to even _block_ here in order to have a complete check on the bundles. b) I don't see the database operations here are heavy/special, every transaction validation involves the database and runtime API call.

Reply:
Well, whether database operations are heavy/special or not, it is good to remove them completely from this check though.

The code you propose here does reduce a few locs by combining the bundle set maintenance and sharing into one, but I prefer the current implementation where the logic of maintenance and sharing are separated because it improves the readability and maintainability of both maintenance and sharing.

Reply:
>it improves the readability and maintainability of both maintenance and sharing.

I guess that's the problem :P, my opinion is on the opposite, I don't see the evidence of performance benefits and it makes the implementation over complicated unnecessarily, which is harmful to the readability and maintainability.

Reply:
Seems we got an opposing view :)

Reply:
The maintenance is about reading/updating the bundle set, when combining maintenance and sharing into one, the bundle set is behind a `Mutex`. 

And when comes to `Mutex`, people reading or changing the code need to think about how long the lock gonna hold, whether there will be a deadlock, will the lock conflict be frequent, and how to ensure the update to bundle set is atomic, etc. And this kind of burden is why I think it hurt readability and maintainability.

When maintenance and sharing are separated, this burden is minimized because the sharing part is scoped into a few lines of very simple code. And the maintenance part just needs to focus on how to reading/updating the bundle set when a new best block is imported. And this is the reason I think it improves readability and maintainability.

Reply:
It's kind of over-engineering from what I see, there are a lot of Mutex usages in our codebase, but I haven't seen a similar pattern.

Reply:
Well, in my point of view, both the current implementation and your proposal work, the discussion here is just about some subtle personal coding style and preference.

And If we have to have a discussion like this to get a consensus, then the abovementioned is what backs my opinion, but that doesn't mean every time I have to spend a lot of time thinking to get something like this implemented because it is just natural to me.

And I do believe it does not bring complexity and has the abovementioned pros, and why not why have to change to another way? If it is indeed important and worth the time then let's discuss more the pros & cons of different preferences to get a consensus.

Reply:
My suggestion was from a standpoint that one general principle is to prefer easy simple understandable code and try to keep consistency with the existing code if possible. The pros you mentioned were not convincing to me for this use case, but I'm not a dictator here. It's absolutely fair to have a personal coding preference, but I can't silently approve the code which looks alien to the existing ones without a decent reason. Could you briefly make a conclusion of our discussion (I guess the main point of your arguments is Mutex) here and post it on slack to get more people's opinions?

Reply:
TL;DR: I'm always in favor of misuse-resistant code, see my understanding and proposal below.

I agree `bundle_stored_in_last_k` is not pretty, but I'm not sure removing it will be a good idea, locks will indeed have to be held for quite some time.

I'm not sure if that is a problem though, it depends on whether there is a concurrent thing happening that will be negatively affected by this. Also the way it is implemented right now is NOT atomic in logical sense because you can plausibly have two `BundleCollector`s that each have their own `bundle_stored_in_last_k` and will basically step on each other toes overriding contents of `bundle_syncer` with conflicting values because their internal logic is NOT actually based on `bundle_syncer`'s contents.

Even if we need to have two of these, I think there should be be a data structure that holds both fields behind mutexes in there and ensures invariants are preserved and we never get a diverging view like I explained in above paragraph. This can be achieved by doing updates in a callback or returning something similar to mutex guard that'll call `bundle_syncer.update()` internally on drop.

That way most of the time quick lock will be taken on `bundle_syncer` and sometimes longer lock will be taken on `bundle_stored_in_last_k`, so performance generally will be identical to current implementation with additional security guarantees. As to deadlocks, you can do `bundle_stored_in_last_k.try_lock()` if you don't expect it to every be accessed concurrently and return an error if it is already held to signal a bug in business logic.

Reply:
> you can plausibly have two BundleCollectors that each have their own bundle_stored_in_last_k and will basically step on each other toes overriding contents of bundle_syncer with conflicting values

Currently, there should only be one `BundleCollector` when the node is running. And because `BundleCollector` didn't `#[derive(Clone)]` if there is more than one `BundleCollector` somehow then they are created by `BundleCollector::new` thus they will not share the same `bundle_syncer`.

We can also make the code more robust by `impl Clone for BundleCollector` and ensure the cloned `BundleCollector` will have a new and empty state though, which prevents people from accidentally `#[derive(Clone)]` for `BundleCollector` get the abovementioned problem.

Reply:
It is ideal when the wrong code doesn't compile. My suggestion is to make it so that that particular issue is non-issue. Manually implementing `Clone` has no impact on ability to use derive later.

Reply:
Manually implementing `Clone` will make `#[derive(Clone)]` fail to compile due to `conflicting implementations of trait Clone`. And unless someone intentionally changes the code to make two `BundleCollector` sharing the same `bundle_syncer` I think it will be okay.

Reply:
It fixes nothing about the API concern I have about `.update()` that should be called with the whole contents.

What do you think about wrapper data structure that encapsulates both though? To the outside it'll look like one data structure, but it'll be more efficient.

Reply:
Okay, I think I got your concern about the `.update()` and the diverging view. I had written a little prototype of your proposal, and it looks good to me as it also fulfills what I need, will push it here after some polish and see how it goes.